File: src/2670.find-the-distinct-difference-array.2777/solution.ts
function distinctDifferenceArray (nums: number[]): number[] {
    const set = new Set<number>();
    const map2 = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        map2.set(nums[i], map2.has(nums[i]) ? map2.get(nums[i])! + 1 : 1);
    }
    const result = new Array<number>(nums.length);
    for (let i = 0; i < nums.length; i++) {
        set.add(nums[i]);
        const c = map2.get(nums[i])!;
        if (c === 1) {
            map2.delete(nums[i]);
        } else {
            map2.set(nums[i], c - 1);
        }

        result[i] = set.size - map2.size;
    }

    return result;
}

File: src/2670.find-the-distinct-difference-array.2777/solution.java
class Solution {
    public int[] distinctDifferenceArray(int[] nums) {
        int[] result = new int[nums.length];
        Set<Integer> set = new HashSet<>();
        int[] suffixs = new int[nums.length];
        for (int i = nums.length - 1; i > 0; i--) {
            set.add(nums[i]);
            suffixs[i] = set.size();
        }

        set.clear();

        for (int i = 0; i < nums.length - 1; i++) {
            set.add(nums[i]);
            result[i] = set.size() - suffixs[i + 1];
        }
        set.add(nums[nums.length - 1]);
        result[nums.length - 1] = set.size();

        return result;
    }
}